### PR TITLE
Add ported Test-Connection cmdlet

### PR DESCRIPTION
**Pull request by iSazonov**
## PR Summary

The Test-Connection cmdlet works on Windows and Unix.

### Implemented:
- Ping
- Continues Ping
- Traceroute
- Detect MTU size (seems don't works on Unix because .Net Core issue)
- Connect to TCP port

### Not jet implemented:
I'm putting this off for the future.
- Detect blackhole routers
- PingPath

### Additional considerations

As you can see in the tests .Net Core has issues in API implementation. As a result, some tests are skipped on Unix. There is also one issue on Windows. Related comments added to tests.
I'm going to open an issue(s) in CoreFX repo.

I think we should break the feedback into two parts:
1. scripting functionality
2. interactive functionality (display output)

Now I have implemented the output to the screen as progress bar and as text (without ETS) - we have to decide what is best to use.

Also we could do a merge if scripting functionality is good and wait feedback for interactive functionality (display output).
I am waiting feedback how we better make the output to the screen for all options.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by iSazonov** [comment]
I really don't know how make reliable CI tests.

**Comment by iSazonov** [comment]
@SteveL-MSFT @TravisEz13 Can we have IPv6 on CIs to pass the tests?

**Comment by TravisEz13** [comment]
On windows, you can detect if the machine has an IPv6 address and run the test or skip it.

**Comment by iSazonov** [comment]
I believe we shouldn't skip any tests and we should pass the same test set on all platforms. The tests is passed locally. Is it acceptable pending the IPv6 tests on CIs? Or remove all the IPv6 tests and leave only IPv4 test?
IPv6 protocol is enabled on Windows by default (CI AppVeyor). It seems we could enable IPv6 on CI Travis too. We can pass the tests without real IPv6 addresses.

**Comment by TravisEz13** [comment]
If you think you can enable IPv6 on travis-ci, feel free to add it to the PR.  I think I miss understood the question.

**Comment by iSazonov** [comment]
@TravisEz13 Thanks! I discover that CI Travis has problems with IPv6. I push new commit with workaround https://github.com/travis-ci/travis-ci/issues/8361#issuecomment-350497804

**Comment by iSazonov** [comment]
The workaround don't help 😕 They removed [localhost alias for ::1](https://github.com/travis-ci/travis-ci/issues/8891).
I can make the tests more stable on CI Appveyor but it doesn't seem to make sense to fight this on CI Travis since they can't do IPv6 for many years. 

I suggest skip (or pending) the test on CI Travis and enable them local and on CI Appveyor.

**Comment by FriedCircuits** [comment]
Would be great to get this merged. We use test-connection to check VPN connections.

**Comment by TravisEz13** [comment]
@FriedCircuits The tests are failing.  We need the tests to be fixed before we can merge.

**Comment by TravisEz13** [comment]
Updated PR description with current Template

**Comment by FriedCircuits** [comment]
@TravisEz13 for sure, I saw the bot marked it as stale.

**Comment by TravisEz13** [comment]
@FriedCircuits Yup, Please feel free to comment to keep a PR from being closed.  But eventually, someone needs to be interested in continuing the work.  An open PR takes some time to track and manage, although we are finding tools to help with that.

**Comment by iSazonov** [comment]
I don't know how fix tests because of IPV6. I could  re-write tests for only IPv4.

**Comment by SteveL-MSFT** [comment]
@iSazonov perhaps we should update the PR for ipv4 and open an issue to add ipv6 tests later

**Comment by iSazonov** [comment]
@TravisEz13 @SteveL-MSFT The PR is ready for review.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 1]
```diff
@@ -0,0 +1,937 @@
...
+
+        // Random value for WriteProgress Activity Id.
+        private static readonly int s_ProgressId = 174593053;
+
+        // Empty message string for Progress Bar.
+        private const string ProgressRecordSpace = " ";
+
+        private const string TestConnectionExceptionId = "TestConnectionException";
+   }
+}
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: missing newline

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs [hunk 2]
```diff
@@ -0,0 +1,937 @@
...
+                // Delay between ping but not after last ping.
+                if (i < Count && Delay > 0)
+                {
+                    Thread.Sleep(delay);
+                }
+            }
+
+            if (!Continues.IsPresent)
+            {
+                WritePingFooter();
```
Comment (TravisEz13) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: maybe move this to a finally and always do this?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        It "BufferSize Should -Be between 0 and 65500" {
+            { Test-Connection $targetName -BufferSize 0 }     | Should Not Throw
+            { Test-Connection $targetName -BufferSize -1 }    | ShouldBeErrorId "ParameterArgumentValidationError,Microsoft.PowerShell.Commands.TestConnectionCommand"
+            { Test-Connection $targetName -BufferSize 65501 } | ShouldBeErrorId "ParameterArgumentValidationError,Microsoft.PowerShell.Commands.TestConnectionCommand"
+        }
+
+        It "BufferSize works" {
+            $result = Test-Connection $targetName -Count 1 -BufferSize 2
+
+            if (!$isWindows) {
```
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1]: Can you add a comment about why we have the special case?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            $WebListener = Start-WebListener
+        }
+        
+        It "Test connection to local host port 80" {
+            Test-Connection '127.0.0.1' -TCPPort $WebListener.HttpPort | Should -BeTrue
+        }
+
+        It "Test connection to unreachable host port 80" {
+            Test-Connection $UnreachableAddress -TCPPort 80 -TimeOut 1 | Should -BeFalse
+        }
```
Comment (TravisEz13) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1]: doesn't look like this is indented properly

**Review by TravisEz13 — CHANGES_REQUESTED** [review]
see comments

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs (hunk 1 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: Fixed.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs]: I think we should break the feedback into two parts:
1. scripting functionality
2. interactive functionality (display output)

Now I have implemented the output to the screen as progress bar and as text (without ETS) - we have to decide what is best to use.

Also we could do a merge if scripting functionality is good and wait feedback for interactive functionality (display output).

As for the footer comment I haven't personal preference. I am waiting feedback how we better make the output to the screen for all options.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1 (hunk 3 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1]: The comment is in Line 44 - should we duplicate it on every `if (!$isWindows) { `?

File: test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1 (hunk 4 above)
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1]: Fixed.

**Review by TravisEz13 — APPROVED** [review]
I still think it would be good to complete the progress.
Explaining special cases in just one test case is not as useful as explaining it in each as often you have to review an individual case.

**Comment by enekux** [comment]
There is a workaround here. You can use the .Net Ping Class instead and until the Test-Connection cmdlet is ported.
More info here:
http://ilovepowershell.com/2010/04/13/powershell-how-to-ping-computers/

**Comment by iSazonov** [comment]
@enekux It has _already_ been ported - you can get it in latest builds. We are waiting feedbacks.

**Pull request merged**